controller_manager/test/test_hardware_spawner.cpp

```diff
@@ -0,0 +1,227 @@
+// limitations under the License.
+
+#include <gtest/gtest.h>
+
+#include <cstdlib>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "controller_manager/controller_manager.hpp"
+#include "controller_manager_test_common.hpp"
+#include "lifecycle_msgs/msg/state.hpp"
+#include "test_chainable_controller/test_chainable_controller.hpp"
+#include "test_controller/test_controller.hpp"
+
+using ::testing::_;
+using ::testing::Return;
+
+using namespace std::chrono_literals;
+class TestHardwareSpawner : public ControllerManagerFixture<controller_manager::ControllerManager>
+{
+public:
+  TestHardwareSpawner() : ControllerManagerFixture<controller_manager::ControllerManager>()
+  {
+    cm_->set_parameter(
+      rclcpp::Parameter("hardware_components_initial_state.unconfigured", "TestSystemHardware"));
+  }
+
+  void SetUp() override
+  {
+    ControllerManagerFixture::SetUp();
+
+    update_executor_ =
+      std::make_shared<rclcpp::executors::MultiThreadedExecutor>(rclcpp::ExecutorOptions(), 2);
+
+    update_executor_->add_node(cm_);
+    update_executor_spin_future_ =
+      std::async(std::launch::async, [this]() -> void { update_executor_->spin(); });
+    // This sleep is needed to prevent a too fast test from ending before the
+    // executor has began to spin, which causes it to hang
+    std::this_thread::sleep_for(50ms);
+  }
+
+  void TearDown() override { update_executor_->cancel(); }
+
+protected:
+  // Using a MultiThreadedExecutor so we can call update on a separate thread from service callbacks
+  std::shared_ptr<rclcpp::Executor> update_executor_;
+  std::future<void> update_executor_spin_future_;
+};
+
+int call_spawner(const std::string extra_args)
+{
+  std::string spawner_script =
+    "python3 -m coverage run --append --branch $(ros2 pkg prefix "
+    "controller_manager)/lib/controller_manager/hardware_spawner ";
+  return std::system((spawner_script + extra_args).c_str());
+}
+
+TEST_F(TestHardwareSpawner, spawner_with_no_arguments_errors)
+{
+  EXPECT_NE(call_spawner(""), 0) << "Missing mandatory arguments";
+}
+
+TEST_F(TestHardwareSpawner, spawner_without_manager_errors_with_given_timeout)
+{
+  EXPECT_NE(call_spawner("TestSystemHardware --controller-manager-timeout 1.0"), 0)
+    << "Wrong controller manager name";
+}
+
+TEST_F(TestHardwareSpawner, spawner_without_component_name_argument)
+{
+  EXPECT_NE(call_spawner("-c test_controller_manager"), 0)
+    << "Missing component name argument parameter";
+}
+
+TEST_F(TestHardwareSpawner, spawner_non_exising_hardware_component)
+{
+  EXPECT_NE(call_spawner("TestSystemHardware1 -c test_controller_manager"), 0)
+    << "Missing component name argument parameter";
+}
+
+TEST_F(TestHardwareSpawner, set_component_to_configured_state_and_back_to_activated)
+{
+  EXPECT_EQ(call_spawner("TestSystemHardware --configure -c test_controller_manager"), 0);
+
+  EXPECT_EQ(call_spawner("TestSystemHardware --activate -c test_controller_manager"), 0);
+}
```

Comment:
This makes sure that the spawner is called properly and exited. Can you check the list_hardware_components and their state after every call, so we can be sure that the component has transitioned properly from the CM